crates/driver/src/domain/competition/solution/encoding.rs

```diff
@@ -7,3 +22,330 @@ pub enum Strategy {
     /// Use logic from this module for encoding
     Domain,
 }
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("invalid interaction: {0:?}")]
+    InvalidInteractionExecution(competition::solution::interaction::Liquidity),
+    #[error("invalid clearing price: {0:?}")]
+    InvalidClearingPrice(eth::TokenAddress),
+    #[error(transparent)]
+    Math(#[from] Math),
+}
+
+pub fn tx(
+    auction_id: competition::auction::Id,
+    solution: &super::Solution,
+    contract: &contracts::GPv2Settlement,
+    approvals: impl Iterator<Item = eth::allowance::Approval>,
+    internalization: settlement::Internalization,
+) -> Result<eth::Tx, Error> {
+    let mut tokens = Vec::new();
+    let mut clearing_prices = Vec::new();
+    let mut trades: Vec<Trade> = Vec::new();
+    let mut pre_interactions = Vec::new();
+    let mut interactions = Vec::new();
+    let mut post_interactions = Vec::new();
+
+    // Encode uniform clearing price vector
+    for (token, price) in solution.prices.clone() {
+        tokens.push(token.0.into());
+        clearing_prices.push(price);
+    }
+
+    // Encode trades with custom clearing prices
+    for trade in solution.trades() {
+        match trade {
+            super::Trade::Fulfillment(trade) => {
```

Comment:
nit: I think I would slightly restructure the code to first compute all the relevant information (i.e. trade object and custom prices) and at the end group all the function calls that push these values into the tokens, trades etc.
My reasoning is that for example tokens and prices have to be pushed in the same order and this is easier to verify if all the push operations are grouped together.
The comment to being very deliberate with the code pushing stuff into the final vectors applies to other functions as well.